providers/okta/okta_auth.go

```diff
@@ -0,0 +1,205 @@
+package okta
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/okta/okta-sdk-golang/v2/okta"
+)
+
+type jsonHAL map[string]json.RawMessage
+
+func (j jsonHAL) Get(key string, dest interface{}) error {
+	return json.Unmarshal(j[key], dest)
+}
+
+// AuthClient encapsulates authentication operations that are not exposed by the regular Okta client
+type AuthClient struct {
+	url url.URL
+	rt  RequestExecutor
+}
+
+type RequestExecutor interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type AuthRequest struct {
+	Audience string `json:"audience,omitempty"`
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Options  struct {
+		MultiOptionalFactorEnroll bool `json:"multiOptionalFactorEnroll"`
+		WarnBeforePasswordExpired bool `json:"warnBeforePasswordExpired"`
+	} `json:"options"`
+}
+
+type AuthResponse struct {
+	Status     string
+	ExpiresAt  time.Time
+	StateToken StateToken
+	Embedded   jsonHAL `json:"_embedded"`
+	Links      jsonHAL `json:"_links"`
+}
+
+func (p AuthResponse) Factors() []okta.UserFactor {
+	var factors []okta.UserFactor
+	p.Embedded.Get("factors", &factors)
+	return factors
+}
+
+func (p AuthResponse) UserID() string {
+	var user struct {
+		ID string `json:"id"`
+	}
+
+	// Error intentionally ignored.
```

Comment:
Can you clarify the intention?

Reply:
I think when I wrote this two years ago, it was ignored because I didn't want to return an error from `UserID()`.

I'll update this